Reserve vectors in FunctionSchema::cloneWithRealTypes

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #151682
* #151697
* #151630
* #151629
* #151628
* __->__ #151627
* #151626

1) reserving is much better than not reserving
2) std::transform for a 1-line-body loop is generally not considered to be an improvement (and doesn't get seem to get boiled away by clang under -Oz)

Differential Revision: [D73013363](https://our.internmc.facebook.com/intern/diff/D73013363/)

swolchok committed:
Reserve vectors in FunctionSchema::cloneWithRealTypes

1) reserving is much better than not reserving
2) std::transform for a 1-line-body loop is generally not considered to be an improvement (and doesn't get boiled away by at least one compiler)

Internal: said compiler is just Apple Clang with -Oz

Differential Revision: [D73013363](https://our.internmc.facebook.com/intern/diff/D73013363/)

[ghstack-poisoned]

Skylion007 commented:
Std::transform also does not preallocate

malfet commented on aten/src/ATen/core/function_schema.cpp:
```diff
@@ -41,9 +41,15 @@ FunctionSchema FunctionSchema::cloneWithRealTypes(bool with_symint) const {
   };
   std::vector<Argument> new_arguments, new_returns;
-  std::transform(arguments().begin(), arguments().end(), std::back_inserter(new_arguments), cloneWithRealTypes);
+  new_arguments.reserve(arguments().size());
```
Mind adding a comment here? Also, if size of the vector is know ahead of time, why not pre-allocate vector to given size on line 43 and just add elements at index?

swolchok replied:
resize() will default-construct all the elements of the vector, which matters for a type like Argument with a non-trivial default constructor.

swolchok replied:
> adding a comment

I don't think there is anything to be commented; we should *always* pre-allocate before building vectors if we know their size.

malfet commented on aten/src/ATen/core/function_schema.cpp:
```diff
@@ -41,9 +41,15 @@ FunctionSchema FunctionSchema::cloneWithRealTypes(bool with_symint) const {
-  std::transform(arguments().begin(), arguments().end(), std::back_inserter(new_arguments), cloneWithRealTypes);
+  new_arguments.reserve(arguments().size());
+  for (const auto& arg: arguments()) {
+    new_arguments.push_back(cloneWithRealTypes(arg));
```
Is there are reason to prefer `push_back` to `emplace_back`
```suggestion
    new_arguments.emplace_back(cloneWithRealTypes(arg));
```

swolchok replied:
push_back has (since C++11) an overload that takes T&&. emplace_back does more complicated shenanigans with forwarding its arguments to the constructor of T. here we have a T&&, and so push_back matches our needs.

swolchok replied:
see also https://quuxplusone.github.io/blog/2021/03/03/push-back-emplace-back/

swolchok committed:
Update on "Reserve vectors in FunctionSchema::cloneWithRealTypes"


1) reserving is much better than not reserving
2) std::transform for a 1-line-body loop is generally not considered to be an improvement (and doesn't get seem to get boiled away by clang under -Oz)

Differential Revision: [D73013363](https://our.internmc.facebook.com/intern/diff/D73013363/)

[ghstack-poisoned]
